# The Hello World of Deep Learning with Neural Networks

### Course 1 - Part 2 - Lesson 2 - Notebook
Like every first app you should start with something super simple that shows the overall scaffolding for how your code works. 

In the case of creating neural networks, the sample I like to use is one where it learns the relationship between two numbers. So, for example, if you were writing code for a function like this, you already know the 'rules' — 


```
float hw_function(float x){
    float y = (2 * x) - 1;
    return y;
}
```

So how would you train a neural network to do the equivalent task? Using data! By feeding it with a set of Xs, and a set of Ys, it should be able to figure out the relationship between them. 

This is obviously a very different paradigm than what you might be used to, so let's step through it piece by piece.


## Imports

Let's start with our imports. Here we are importing TensorFlow and calling it tf for ease of use.

We then import a library called numpy, which helps us to represent our data as lists easily and quickly.

The framework for defining a neural network as a set of Sequential layers is called keras, so we import that too.

In [1]:
import tensorflow as tf
import numpy as np
from tensorflow import keras

## Define and Compile the Neural Network

Next we will create the simplest possible neural network. It has 1 layer, and that layer has 1 neuron, and the input shape to it is just 1 value.

In [2]:
model = tf.keras.Sequential([keras.layers.Dense(units=1, input_shape=[1])])

Now we compile our Neural Network. When we do so, we have to specify 2 functions, a loss and an optimizer.

If you've seen lots of math for machine learning, here's where it's usually used, but in this case it's nicely encapsulated in functions for you. But what happens here — let's explain...

We know that in our function, the relationship between the numbers is y=2x-1. 

When the computer is trying to 'learn' that, it makes a guess...maybe y=10x+10. The LOSS function measures the guessed answers against the known correct answers and measures how well or how badly it did.

It then uses the OPTIMIZER function to make another guess. Based on how the loss function went, it will try to minimize the loss. At that point maybe it will come up with somehting like y=5x+5, which, while still pretty bad, is closer to the correct result (i.e. the loss is lower)

It will repeat this for the number of EPOCHS which you will see shortly. But first, here's how we tell it to use 'MEAN SQUARED ERROR' for the loss and 'STOCHASTIC GRADIENT DESCENT' for the optimizer. You don't need to understand the math for these yet, but you can see that they work! :)

Over time you will learn the different and appropriate loss and optimizer functions for different scenarios. 


In [3]:
model.compile(optimizer='sgd', loss='mean_squared_error')

## Providing the Data

Next up we'll feed in some data. In this case we are taking 6 xs and 6ys. You can see that the relationship between these is that y=2x-1, so where x = -1, y=-3 etc. etc. 

A python library called 'Numpy' provides lots of array type data structures that are a defacto standard way of doing it. We declare that we want to use these by specifying the values as an np.array[]

In [4]:
xs = np.array([-1.0,  0.0, 1.0, 2.0, 3.0, 4.0], dtype=float)
ys = np.array([-3.0, -1.0, 1.0, 3.0, 5.0, 7.0], dtype=float)

# Training the Neural Network

The process of training the neural network, where it 'learns' the relationship between the Xs and Ys is in the **model.fit**  call. This is where it will go through the loop we spoke about above, making a guess, measuring how good or bad it is (aka the loss), using the opimizer to make another guess etc. It will do it for the number of epochs you specify. When you run this code, you'll see the loss on the right hand side.

In [5]:
model.fit(xs, ys, epochs=500)

Epoch 1/500
6/6 [==============================] - 1s 155ms/step - loss: 44.8001
Epoch 2/500
6/6 [==============================] - 0s 265us/step - loss: 35.6206
Epoch 3/500
6/6 [==============================] - 0s 259us/step - loss: 28.3909
Epoch 4/500
6/6 [==============================] - 0s 229us/step - loss: 22.6954
Epoch 5/500
6/6 [==============================] - 0s 277us/step - loss: 18.2071
Epoch 6/500
6/6 [==============================] - 0s 214us/step - loss: 14.6687
Epoch 7/500
6/6 [==============================] - 0s 219us/step - loss: 11.8777
Epoch 8/500
6/6 [==============================] - 0s 210us/step - loss: 9.6750
Epoch 9/500
6/6 [==============================] - 0s 213us/step - loss: 7.9353
Epoch 10/500
6/6 [==============================] - 0s 219us/step - loss: 6.5599
Epoch 11/500
6/6 [==============================] - 0s 218us/step - loss: 5.4713
Epoch 12/500
6/6 [==============================] - 0s 251us/step - loss: 4.6085
Epoch 13/500
6/6 [============

Epoch 103/500
6/6 [==============================] - 0s 227us/step - loss: 0.2337
Epoch 104/500
6/6 [==============================] - 0s 220us/step - loss: 0.2289
Epoch 105/500
6/6 [==============================] - 0s 224us/step - loss: 0.2242
Epoch 106/500
6/6 [==============================] - 0s 238us/step - loss: 0.2196
Epoch 107/500
6/6 [==============================] - 0s 263us/step - loss: 0.2151
Epoch 108/500
6/6 [==============================] - 0s 300us/step - loss: 0.2106
Epoch 109/500
6/6 [==============================] - 0s 235us/step - loss: 0.2063
Epoch 110/500
6/6 [==============================] - 0s 228us/step - loss: 0.2021
Epoch 111/500
6/6 [==============================] - 0s 226us/step - loss: 0.1979
Epoch 112/500
6/6 [==============================] - 0s 248us/step - loss: 0.1939
Epoch 113/500
6/6 [==============================] - 0s 289us/step - loss: 0.1899
Epoch 114/500
6/6 [==============================] - 0s 284us/step - loss: 0.1860
Epoch 115/500
6/

Epoch 203/500
6/6 [==============================] - 0s 243us/step - loss: 0.0293
Epoch 204/500
6/6 [==============================] - 0s 238us/step - loss: 0.0287
Epoch 205/500
6/6 [==============================] - 0s 235us/step - loss: 0.0281
Epoch 206/500
6/6 [==============================] - 0s 226us/step - loss: 0.0276
Epoch 207/500
6/6 [==============================] - 0s 215us/step - loss: 0.0270
Epoch 208/500
6/6 [==============================] - 0s 230us/step - loss: 0.0264
Epoch 209/500
6/6 [==============================] - 0s 286us/step - loss: 0.0259
Epoch 210/500
6/6 [==============================] - 0s 244us/step - loss: 0.0254
Epoch 211/500
6/6 [==============================] - 0s 262us/step - loss: 0.0248
Epoch 212/500
6/6 [==============================] - 0s 244us/step - loss: 0.0243
Epoch 213/500
6/6 [==============================] - 0s 230us/step - loss: 0.0238
Epoch 214/500
6/6 [==============================] - 0s 214us/step - loss: 0.0233
Epoch 215/500
6/

Epoch 303/500
6/6 [==============================] - 0s 237us/step - loss: 0.0037
Epoch 304/500
6/6 [==============================] - 0s 215us/step - loss: 0.0036
Epoch 305/500
6/6 [==============================] - 0s 237us/step - loss: 0.0035
Epoch 306/500
6/6 [==============================] - 0s 226us/step - loss: 0.0035
Epoch 307/500
6/6 [==============================] - 0s 216us/step - loss: 0.0034
Epoch 308/500
6/6 [==============================] - 0s 239us/step - loss: 0.0033
Epoch 309/500
6/6 [==============================] - 0s 223us/step - loss: 0.0032
Epoch 310/500
6/6 [==============================] - 0s 246us/step - loss: 0.0032
Epoch 311/500
6/6 [==============================] - 0s 220us/step - loss: 0.0031
Epoch 312/500
6/6 [==============================] - 0s 212us/step - loss: 0.0031
Epoch 313/500
6/6 [==============================] - 0s 244us/step - loss: 0.0030
Epoch 314/500
6/6 [==============================] - 0s 294us/step - loss: 0.0029
Epoch 315/500
6/

Epoch 402/500
6/6 [==============================] - 0s 244us/step - loss: 4.7157e-04
Epoch 403/500
6/6 [==============================] - 0s 214us/step - loss: 4.6188e-04
Epoch 404/500
6/6 [==============================] - 0s 247us/step - loss: 4.5239e-04
Epoch 405/500
6/6 [==============================] - 0s 206us/step - loss: 4.4310e-04
Epoch 406/500
6/6 [==============================] - 0s 226us/step - loss: 4.3400e-04
Epoch 407/500
6/6 [==============================] - 0s 243us/step - loss: 4.2508e-04
Epoch 408/500
6/6 [==============================] - 0s 232us/step - loss: 4.1635e-04
Epoch 409/500
6/6 [==============================] - 0s 228us/step - loss: 4.0780e-04
Epoch 410/500
6/6 [==============================] - 0s 235us/step - loss: 3.9943e-04
Epoch 411/500
6/6 [==============================] - 0s 231us/step - loss: 3.9122e-04
Epoch 412/500
6/6 [==============================] - 0s 234us/step - loss: 3.8319e-04
Epoch 413/500
6/6 [==============================] - 0

Epoch 498/500
6/6 [==============================] - 0s 230us/step - loss: 6.4303e-05
Epoch 499/500
6/6 [==============================] - 0s 233us/step - loss: 6.2983e-05
Epoch 500/500
6/6 [==============================] - 0s 229us/step - loss: 6.1689e-05


Ok, now you have a model that has been trained to learn the relationshop between X and Y. You can use the **model.predict** method to have it figure out the Y for a previously unknown X. So, for example, if X = 10, what do you think Y will be? Take a guess before you run this code:

In [6]:
print(model.predict([10.0]))

[[18.977085]]


You might have thought 19, right? But it ended up being a little under. Why do you think that is? 

Remember that neural networks deal with probabilities, so given the data that we fed the NN with, it calculated that there is a very high probability that the relationship between X and Y is Y=2X-1, but with only 6 data points we can't know for sure. As a result, the result for 10 is very close to 19, but not necessarily 19. 

As you work with neural networks, you'll see this pattern recurring. You will almost always deal with probabilities, not certainties, and will do a little bit of coding to figure out what the result is based on the probabilities, particularly when it comes to classification.

To determine how the predictions are made, we can look at the internal variables of the Dense layer. 

In [7]:
print("These are the weights used: {}".format(model.get_weights()))

These are the layer variables: [array([[1.996679]], dtype=float32), array([-0.9897032], dtype=float32)]


Note that the weights 1.9966 and -0.9897 are close to 2x -1. Now let's try more neurons. Here we use a layer with 3 neurons (l0) and an output layer (l1). 

In [43]:
l0 = tf.keras.layers.Dense(units=3, input_shape=[1])  # layer with 3 neurons
l1 = tf.keras.layers.Dense(units=1)                   # output layer
model2 = tf.keras.Sequential([l0, l1])                
model2.compile(optimizer='sgd', loss='mean_squared_error')
model2.fit(xs, ys, epochs=500)                        # training the model
print("The prediction for 10 is {}".format(model2.predict([10.0])))   # predicting the value of 10

Epoch 1/500
6/6 [==============================] - 0s 33ms/step - loss: 23.6862
Epoch 2/500
6/6 [==============================] - 0s 252us/step - loss: 13.8995
Epoch 3/500
6/6 [==============================] - 0s 248us/step - loss: 8.6330
Epoch 4/500
6/6 [==============================] - 0s 264us/step - loss: 5.3034
Epoch 5/500
6/6 [==============================] - 0s 250us/step - loss: 3.2171
Epoch 6/500
6/6 [==============================] - 0s 290us/step - loss: 2.0108
Epoch 7/500
6/6 [==============================] - 0s 243us/step - loss: 1.3770
Epoch 8/500
6/6 [==============================] - 0s 412us/step - loss: 1.0657
Epoch 9/500
6/6 [==============================] - 0s 247us/step - loss: 0.9119
Epoch 10/500
6/6 [==============================] - 0s 248us/step - loss: 0.8272
Epoch 11/500
6/6 [==============================] - 0s 242us/step - loss: 0.7710
Epoch 12/500
6/6 [==============================] - 0s 238us/step - loss: 0.7265
Epoch 13/500
6/6 [==================

Epoch 103/500
6/6 [==============================] - 0s 258us/step - loss: 0.0022
Epoch 104/500
6/6 [==============================] - 0s 255us/step - loss: 0.0020
Epoch 105/500
6/6 [==============================] - 0s 236us/step - loss: 0.0019
Epoch 106/500
6/6 [==============================] - 0s 237us/step - loss: 0.0018
Epoch 107/500
6/6 [==============================] - 0s 258us/step - loss: 0.0017
Epoch 108/500
6/6 [==============================] - 0s 311us/step - loss: 0.0015
Epoch 109/500
6/6 [==============================] - 0s 312us/step - loss: 0.0014
Epoch 110/500
6/6 [==============================] - 0s 253us/step - loss: 0.0013
Epoch 111/500
6/6 [==============================] - 0s 269us/step - loss: 0.0013
Epoch 112/500
6/6 [==============================] - 0s 237us/step - loss: 0.0012
Epoch 113/500
6/6 [==============================] - 0s 310us/step - loss: 0.0011
Epoch 114/500
6/6 [==============================] - 0s 278us/step - loss: 0.0010
Epoch 115/500
6/

Epoch 199/500
6/6 [==============================] - 0s 236us/step - loss: 2.4555e-06
Epoch 200/500
6/6 [==============================] - 0s 242us/step - loss: 2.2868e-06
Epoch 201/500
6/6 [==============================] - 0s 239us/step - loss: 2.1300e-06
Epoch 202/500
6/6 [==============================] - 0s 251us/step - loss: 1.9837e-06
Epoch 203/500
6/6 [==============================] - 0s 256us/step - loss: 1.8475e-06
Epoch 204/500
6/6 [==============================] - 0s 258us/step - loss: 1.7205e-06
Epoch 205/500
6/6 [==============================] - 0s 253us/step - loss: 1.6023e-06
Epoch 206/500
6/6 [==============================] - 0s 254us/step - loss: 1.4922e-06
Epoch 207/500
6/6 [==============================] - 0s 249us/step - loss: 1.3900e-06
Epoch 208/500
6/6 [==============================] - 0s 254us/step - loss: 1.2944e-06
Epoch 209/500
6/6 [==============================] - 0s 254us/step - loss: 1.2056e-06
Epoch 210/500
6/6 [==============================] - 0

Epoch 295/500
6/6 [==============================] - 0s 252us/step - loss: 2.6560e-09
Epoch 296/500
6/6 [==============================] - 0s 306us/step - loss: 2.4798e-09
Epoch 297/500
6/6 [==============================] - 0s 259us/step - loss: 2.3092e-09
Epoch 298/500
6/6 [==============================] - 0s 250us/step - loss: 2.1472e-09
Epoch 299/500
6/6 [==============================] - 0s 244us/step - loss: 1.9984e-09
Epoch 300/500
6/6 [==============================] - 0s 372us/step - loss: 1.8570e-09
Epoch 301/500
6/6 [==============================] - 0s 289us/step - loss: 1.7256e-09
Epoch 302/500
6/6 [==============================] - 0s 277us/step - loss: 1.6150e-09
Epoch 303/500
6/6 [==============================] - 0s 281us/step - loss: 1.4988e-09
Epoch 304/500
6/6 [==============================] - 0s 283us/step - loss: 1.4051e-09
Epoch 305/500
6/6 [==============================] - 0s 278us/step - loss: 1.3068e-09
Epoch 306/500
6/6 [==============================] - 0

Epoch 391/500
6/6 [==============================] - 0s 244us/step - loss: 4.2402e-12
Epoch 392/500
6/6 [==============================] - 0s 287us/step - loss: 4.1140e-12
Epoch 393/500
6/6 [==============================] - 0s 301us/step - loss: 3.5551e-12
Epoch 394/500
6/6 [==============================] - 0s 255us/step - loss: 3.4959e-12
Epoch 395/500
6/6 [==============================] - 0s 268us/step - loss: 3.3259e-12
Epoch 396/500
6/6 [==============================] - 0s 285us/step - loss: 3.0245e-12
Epoch 397/500
6/6 [==============================] - 0s 292us/step - loss: 2.7433e-12
Epoch 398/500
6/6 [==============================] - 0s 247us/step - loss: 2.7001e-12
Epoch 399/500
6/6 [==============================] - 0s 254us/step - loss: 2.5538e-12
Epoch 400/500
6/6 [==============================] - 0s 265us/step - loss: 2.5130e-12
Epoch 401/500
6/6 [==============================] - 0s 279us/step - loss: 2.3898e-12
Epoch 402/500
6/6 [==============================] - 0

Epoch 487/500
6/6 [==============================] - 0s 295us/step - loss: 9.7818e-13
Epoch 488/500
6/6 [==============================] - 0s 230us/step - loss: 9.7818e-13
Epoch 489/500
6/6 [==============================] - 0s 235us/step - loss: 9.7818e-13
Epoch 490/500
6/6 [==============================] - 0s 300us/step - loss: 9.7818e-13
Epoch 491/500
6/6 [==============================] - 0s 288us/step - loss: 9.7818e-13
Epoch 492/500
6/6 [==============================] - 0s 286us/step - loss: 9.7818e-13
Epoch 493/500
6/6 [==============================] - 0s 289us/step - loss: 9.7818e-13
Epoch 494/500
6/6 [==============================] - 0s 281us/step - loss: 9.7818e-13
Epoch 495/500
6/6 [==============================] - 0s 316us/step - loss: 9.7818e-13
Epoch 496/500
6/6 [==============================] - 0s 309us/step - loss: 9.7818e-13
Epoch 497/500
6/6 [==============================] - 0s 273us/step - loss: 9.7818e-13
Epoch 498/500
6/6 [==============================] - 0

Adding more layers and neurons seem to improve the prediction. And looking at the internal variables of the layers we see a more complicated set of weights as opposed to the near 2x-1 relation with 1 neuron. 

In [15]:
print("These are the weights for the hidden layer with 3 neurons: {}".format(l0.get_weights()))
print("These are the weights for layer1: {}".format(l1.get_weights()))

These are the weights for layer0: [array([[ 1.5153776,  1.106216 , -0.5033325]], dtype=float32), array([-0.4466807, -0.1019366, -0.1317457], dtype=float32)]
These are the weights for layer1: [array([[1.3025775 ],
       [0.23423946],
       [0.46294758]], dtype=float32), array([-0.33329388], dtype=float32)]


In [44]:
input_number=10
l0_sum=input_number*l0.get_weights()[0]+l0.get_weights()[1]
print("Hidden layer value= {}".format(l0_sum))
l1_sum=np.sum(l0_sum*np.transpose(l1.get_weights()[0])) + l1.get_weights()[1]
print("Output layer value= {}".format(l1_sum[0]))

Hidden layer value= [[-0.31648055 12.06235     9.856079  ]]
Output layer value= 18.999998092651367


This shows how the prediction was obtained based on the input (10) and the weights from the trained model. 